daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DFDLXTypeCalcFunctions.scala

```diff
@@ -0,0 +1,239 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dpath
+
+import org.apache.daffodil.xml.QName
+import org.apache.daffodil.util.Maybe
+import org.apache.daffodil.processors.TypeCalculator
+import org.apache.daffodil.processors.SimpleTypeRuntimeData
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.dpath.NodeInfo.PrimType.IntegerKind
+import org.apache.daffodil.dpath.NodeInfo.PrimType.String
+import org.apache.daffodil.dpath.NodeInfo.PrimType.Integer
+
+trait typeCalculatorNamedDispatch { self: FNTwoArgs =>
+  var srcType: NodeInfo.Kind = null
+  var dstType: NodeInfo.Kind = null
+
+  override def validateType(srcTypes: Seq[NodeInfo.Kind], dstType: NodeInfo.Kind): Unit = {
+    //We can't actually verify anything yet, because we need to wait until run() is
+    //called to know what calculator we are using
```

Comment:
I agree with the naming issue. The original idea was just that I was deferring type-checking to expression nodes themselves, which is a layer of abstraction lower then we normally do, so this function was providing a back door to sneak the type information into the nodes without tearing up too much existing code. If there is a pre-existing way of getting this type information at the layer I need it, I can remove this function entirely

The reason I ended up deferring to runtime is that we could, in principle, have expressions like:

{ dfdl:inputTypeCalcInt(../functionName, ../functionArg) }

where the type calculator to be used is determined by the infoset itself. Mike had suggested just not allowing this. If we insist the function name be a constant, we should be able to move the validation back to compile time (along with a handleful of other missing/invalid typeCalculator errors).

Even if someone comes along with a reason to need the dynamisism, it might still be worth it to specialize the constant case to provide SDEs at compile time.

Reply:
In DFDL XPath, everything is strongly typed, so at schema compile time you should know what the type of functionName and functionArg are, regardless if they are variables or constants. My gut feeling is that in Expression.scala where you have things like this:
```scala
      case (RefQName(_, "inputTypeCalcInt", DFDL), args) =>
        FNTwoArgsExpr(functionQNameString, functionQName, args,
          NodeInfo.Int, NodeInfo.String, NodeInfo.AnyAtomic, DFDLXInputTypeCalcInt(_))
```
You could replace FNTwoArgsExpr with a custom Expression, perhaps something that ineherits FNTwoArgsExpr and extends it to provide extra type information when it calls the DFDLXInputTypeCalcInt recipe constrcutor.

In fact, it seems like the whole getCalculator stuff could get determined at compile time. We should know the type of the arguments (because strongly typed) so the calculator should be known at compile time? I think...